### ISSUE TITLE
🐛 BUG: styles not rendering when importing layout from components folder

### ISSUE THREAD

**Issue by waspeer** [withastro/astro#1964]
### What version of `astro` are you using?

0.21.0

### What package manager are you using?

pnpm/npm

### What operating system are you using?

Mac

### Describe the Bug

This is a random one, but I found out that my style tags are not rendering when I'm importing my layout component from a folder called `components` instead of a folder called `layouts`. Since the docs describe the usage of these folder names as conventions instead of rules, I was not expecting different behavior.

So, this does *not* work:

```astro
---
import Layout from '../components/layout.astro';
---

<Layout>
  <h1>Hello</h1>
</Layout>

<style>
  h1 {
    color: red
  }
</style>
```

But when I switch out the import to `import Layout from '../layouts/layout.astro';` it does.

One other nuance I discovered is that wrapping the whole thing inside some tag also seems to resolve the issue:

```astro
---
import Layout from '../components/layout.astro';
---

<div>
  <Layout>
    <h1>Hello</h1> 
  </Layout>
</div>

<style>
  h1 {
    color: red
  }
</style>
```

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-vyy3mj?file=src/pages/index.astro

**Comment by jonathantneal** [comment]
I can confirmed that this is happening when the page does not have a `<body>`.

It _does_ work when the `<body>` exists before the `<h1>`. <sup>[Reproduction](https://stackblitz.com/edit/github-vyy3mj-3uhrfw?file=src/pages/index.astro)</sup>

- If both should work, then I suspect this is a bug.
- If `<body>` must come before or around `<Layout>`, then I suspect this should be [documented in `/guides/styling/`](https://docs.astro.build/guides/styling/).

**Comment by waspeer** [comment]
In the second example in my issue I already pointed out that this is issue resolves when the layout is surrounded by an arbitrary element. This works also for other tags than `body`. The StackBlitz I included in the issue is a very simplified example of my project. In my original project the `body` tags (as well as `head`, `html` etc) are part of my layout.

**Comment by matthewp** [comment]
Thanks, yeah this is a bug then. I think the issue is that we parse something in `layouts` differently but that's really a workaround. I think @natemoo-re worked on that part so maybe he can speak to it.

**Comment by drwpow** [comment]
I’m not sure if this is a bug or not. It’s certainly unexpected, but there is some aspect of Astro behaving as intended.

A major feature of Astro is being able to use `.astro` components to generate parts of the `<head>`. And in the case of layouts, it may also choose to handle the `<html>` tag as well. What’s happening here is because neither `/src/pages/index.astro` nor `/src/components/layout.astro` handles `<html>`, `<head>`, or `<body>`, they both defer to each other and basically the `<head>` winds up missing (taking the styles with it).

As @matthewp said, the `/src/layouts` folder does have “page-like” behavior, and because the first element in there is a body element, Astro correctly figures out that it should be inserting the `<html>`, `<head>`, and `<body>` tags because the first line isn’t another Astro component, but a standard HTML tag.

Also keep in mind our previous compiler sometimes had issues dealing with inserting too many `html/head/body` tags or too few. Our new compiler is much better, but in situations like this where it’s ambiguous where the component should live, it has trouble.

I’d say **best practice** is to explicitly declare your `<head>` and `<body>` somewhere within your page tree, since you’ll more than likely need to manage these at some point. But it seems for now, if the first element is a Component, like others have pointed out, this seems to be required.

**Comment by drwpow** [comment]
🤔 But there does seem to be a bug where if `<head>` or `<body>` is put in the component, the compiler throws an error. That probably shouldn’t happen!

**Comment by waspeer** [comment]
As I mentioned before, this bug is unrelated to the absence of the `html`/`body`/`head` elements. I left them out of my example to be more to the point. In my project I encountered this issue while my layout included all the html boilerplate, like this: 

https://stackblitz.com/edit/github-vyy3mj-ccxepf?file=src/pages/index.astro

This is in line with [the examples shown in the docs](https://docs.astro.build/core-concepts/layouts/#example), with the exception of the layout being in a folder called `components` instead of `layouts`. Since it is mentioned in the docs that [it is not required to put your layout components in a folder called `layouts`](https://docs.astro.build/core-concepts/project-structure/#srclayouts), I believe this is a bug.

It is correct that this issue resolves when the layout is surrounded by an arbitrary element (as seen in the last code example in my original issue), but this works for any element, so this is not exclusive to html boilerplate elements like `html`, `body`, etc.

**Comment by drwpow** [comment]
Update on this: other users have experienced issues with this, so for now we’re encouraging users to keep `<head>` and `<body>` inside either an Astro Page or Astro Layout.

> Since it is mentioned in the docs that [it is not required to put your layout components in a folder called `layouts`](https://docs.astro.build/core-concepts/project-structure/#srclayouts), I believe this is a bug.

This is true! And isn’t changing. But the one caveat for now is still `<head>` and `<body>` tags. You can still create components to manage content in either the head or body, but the start & end tags for now must be in a page or layout.

We’ll be talking this through more, and this isn’t a final decision! Just trying to unblock people for now because there are things to consider (for example, if we allow `<head>` tags within components, what should happen? Should they merge together or should one “win out”?)

**Comment by tony-sull** [comment]
An [RFC](https://github.com/withastro/rfcs/blob/main/proposals/0007-finalize-head-body.md) was accepted not too long ago to finalize `<head>` and `<body>` behavior :+1:

**Comment by natemoo-re** [comment]
This should be fixed in the newly released `astro@0.24.0`!

Please give feel free to reopen this issue if you're still having trouble.
